modules/programs/waybar.nix

```
@@ -242,8 +242,16 @@ in {
         settingsModules =
           optionalAttrs (configuration.modules != { }) configuration.modules;
       in removeNulls (settingsWithoutModules // settingsModules);
+
+    # Allow using attrs for settings instead of a list in order to more easily override
+    settings = with builtins;
+      if typeOf cfg.settings == "set" then
+        attrValues cfg.settings
+      else
+        cfg.settings;
```

```suggestion
    settings =
      if builtins.isAttrs cfg.settings then
        lib.attrValues cfg.settings
      else
        cfg.settings;
```

Maybe we could also do change the settings type to also automatically set the bar name/id :  `types.submodule ({ name ? null, ... }: { options.name = mkOption { default = name; }; })`.

---

Fixed.
Not quite sure what you mean about the settings type though

---

Nevermind about the settings idea, it wouldn't work correctly with `listOf`.

The thing about `types.listOf` and `types.attrsOf` is that if a `types.submodule` or `types.submoduleWith` is used with the function form below, a `name` parameter (for instance the key in the attribute set) is passed to the submodule and can be used (e.g. to set the default value of an option).

Sadly, `listOf` also passes a `name` in the form of `[definition X-entry X]` which is not a proper CSS name, this would break configurations.

(don't apply this patch, this is just for the purpose of explaining the `types.submodule` thing above).
``` diff
diff --git a/modules/programs/waybar.nix b/modules/programs/waybar.nix
index b7d56b66..0b2a74ee 100644
--- a/modules/programs/waybar.nix
+++ b/modules/programs/waybar.nix
@@ -19,7 +19,7 @@ let
     };
 
   waybarBarConfig = with types;
-    submodule {
+    submodule ({ name ? null, ... }: {
       freeformType = jsonFormat.type;
 
       options = {
@@ -125,7 +125,8 @@ let
 
         name = mkOption {
           type = nullOr str;
-          default = null;
+          # `name` inherited from `attrsOf submodule`
+          default = name;
           description =
             "Optional name added as a CSS class, for styling multiple waybars.";
           example = "waybar-1";
@@ -139,7 +140,7 @@ let
             "Option to disable the use of gtk-layer-shell for popups.";
         };
       };
-    };
+    });
 in {
   meta.maintainers = with lib.maintainers; [ berbiche ];
 
diff --git a/tests/modules/programs/waybar/settings-complex-expected.json b/tests/modules/programs/waybar/settings-complex-expected.json
index 87e4f690..d8983e7b 100644
--- a/tests/modules/programs/waybar/settings-complex-expected.json
+++ b/tests/modules/programs/waybar/settings-complex-expected.json
@@ -29,6 +29,7 @@
       "battery#bat2",
       "clock"
     ],
+    "name": "mainBar",
     "output": [
       "DP-1"
     ],
@@ -52,6 +53,7 @@
       "sway/mode"
     ],
     "modules-right": [],
+    "name": "secondaryBar",
     "output": [
       "!DP-1"
     ],
```

---

Ohh, thanks for the explanation. I didn't understant that you were talking about CSS names. It would have been a nice feature indeed !